gui/connected_device.h

```diff
@@ -0,0 +1,39 @@
+#ifndef CONNECTED_DEVICE_H
+#define CONNECTED_DEVICE_H
+
+#include "../common/generated_libs/network_protocol/network_protocol.pb.h"
+#include <functional>
+
+// Represents a connection to the device running the controller.
+class ConnectedDevice {
+public:
+  virtual ~ConnectedDevice() = default;
+
+  // Performs one round of state exchange: sends GuiStatus to the controller
+  // (containing ventilator parameters) and obtains back a ControllerStatus
+  // (containing current sensor values and such).
```

Comment:
This is the part of the PR that I'm tripping over the most.

There's no explicit exchange of one message for one message in the protocol.  Both sides can send messages at their own pace, so long as they meet a minimum cadence (which we haven't defined).

We can pretend that there's an exchange, if we want.  But...what does that buy us?

Reply:
I made it this way because that's what I showed you and Albertas and neither raised an objection :) And because that's what the controller seems to be doing [here](https://github.com/RespiraWorks/VentilatorSoftware/blob/master/controller/src/main.cpp#L67)?

What would be a better protocol? Does the serial port allow asynchronous bidi communication? E.g. if GUI periodically sends its statuses AND in another thread continuously receive statuses from the controller through some blocking API, does that work - or do these things have to wait for each other?

Reply:
> And because that's what the controller seems to be doing here?

Heh.  The difference is that in the controller code, we don't keep history.  So that function means, "get me the latest GuiStatus, and send the current ControllerStatus (if it's time to send)".

Whereas here, every time we call ExchangeState we push the ControllerStatus into a queue.  So I presumed the function meant something quite different, because we don't want to "make up" data points by assuming the ControllerStatus hasn't changed.

> Does the serial port allow asynchronous bidi communication?

That is what I am told.

Note that at least the Arduino controller has a very small rx buffer, something like 64 bytes.  So you'll be unable to send more than that at once.

Reply:
Alright, I split this API into a separate Send and Receive, and am now invoking them both separately periodically in main.cpp + a TODO to fix it more if it isn't right. I guess the rx buffer is something the "real" implementation of this API will have to worry about.